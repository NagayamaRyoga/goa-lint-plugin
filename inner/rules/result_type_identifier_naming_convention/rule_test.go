package result_type_identifier_naming_convention_test

import (
	"log"
	"os"
	"testing"

	"github.com/NagayamaRyoga/goalint/inner/rules/result_type_identifier_naming_convention"
	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/stretchr/testify/assert"
	"goa.design/goa/v3/dsl"
	"goa.design/goa/v3/eval"
	"goa.design/goa/v3/expr"
)

func TestRule(t *testing.T) {
	t.Parallel()

	var (
		resultTypeWithValidID   = dsl.ResultType("application/vnd.user", func() {})
		resultTypeWithInvalidID = dsl.ResultType("book", func() {})
	)

	// given
	err := eval.RunDSL()
	assert.NoError(t, err)

	testCases := []struct {
		description string
		userType    expr.UserType
		wantReports int
	}{
		{
			description: "success",
			userType:    resultTypeWithValidID,
			wantReports: 0,
		},
		{
			description: "failed",
			userType:    resultTypeWithInvalidID,
			wantReports: 1,
		},
	}
	for _, tc := range testCases {
		tc := tc
		t.Run(tc.description, func(t *testing.T) {
			t.Parallel()

			snaps := snaps.WithConfig(snaps.Dir("testdata"))

			logger := log.New(os.Stdout, "", 0)
			cfg := result_type_identifier_naming_convention.NewConfig()
			rule := result_type_identifier_naming_convention.NewRule(logger, cfg)

			// when
			got := rule.WalkResultType(tc.userType)
			assert.Equal(t, tc.wantReports, len(got))
			snaps.MatchSnapshot(t, got.String())
		})
	}
}
