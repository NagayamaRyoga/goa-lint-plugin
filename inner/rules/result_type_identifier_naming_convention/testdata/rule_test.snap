
[TestRule/success - 1]

---

[TestRule/failed - 1]
Found 1 errors and 0 warnings
[ResultTypeIdentifierNamingConvention]: error in ResultType("book"): ResultType identifier should have prefix 'application/vnd.'

---
